clean up some writing in the spec (ch1, ch2, ch11) (#859)
eclipse-ee4j/jpa-api: spec/src/main/asciidoc/ch02-entities.adoc

  via the property accessor methods. Every property not annotated with the
  `Transient` annotation is persistent.

Mapping annotations must not be applied to a field or property marked `static`,

> `transient`, or `@Transient`, since the field or property is not persistent.

Reviewer: ```suggestion
`transient`, or `@jakarta.persistence.Transient`, since such field or property is not persistent.
```

Author: @marko-bekhta See 51ae429 and eab5d4e.

for example, `Set<Order>` is preferred to the raw type `Set`.

[NOTE]
Terminology Note: The terms _collection_ and _collection-valued_ are used
in this specification to denote any of the above types, unless further
qualified. In cases where a `java.util.Collection` type--or one of its
subtypes--must be distinguished, the type is identified explicitly. The
terms _map_ and _map collection_ are used to denote a collection of type
`java.util.Map`.

A collection implementation type such as `HashSet` or `ArrayList` may be
used by the application to initialize a collection-valued field or property
before the entity is made persistent. Once the entity becomes managed
(or detached), subsequent access to the collection must be through the
interface type. The persistence provider runtime is permitted to replace,
at any time, any reference held by an entity to a collection instantiated by
the application with a different provider-specific object implementing the
same interface type.

> ==== Persistent Attribute Type [[categories]]

Author: See fffef34.

Every persistent field or property of an entity is either:

- an attribute of **basic type**, as defined in <<a486>>, or of a type

>   converted to a basic type through use of a _converter_, as defined

Author: Well yeah use of italics there is debatable. The thing is that it's the first time that the term occurs in the specification, and I *do* link to its definition, so I sorta am defining it.

well-defined object identity unique to the instance and is passed
_by reference_, meaning that assignment of the instance to a field,
local variable, or method parameter, followed by later access to
the assigned value, preserves the identity of the instance.

In general, operations performed by a persistence provider are not
required to preserve the object identity of a value stored in an
attribute of an entity. The provider is permitted to modify the
value held in the attribute to a new value with a distinct object
identity. Such modifications should never alter the _persistent_
representation of the entity because the Java object identity of
a value held in an attribute is never part of the persistent
state of the entity.

In particular:

- A basic type, an embeddable class, or any collection type is
  considered a _value type_, and the provider is not required
  to respect the integrity of shared references to an instance
  of such a type. If multiple fields referenced the same instance
  of a value type when an entity was made persistent, those fields
  hold references to distinct instances of the value type when the

>   entity is reloaded from the database.

Reviewer: ```suggestion
  of such a type. If multiple fields referenced the same instance
  of a value type when an entity was made persistent, those fields
  may hold references to distinct instances of the value type when 
  the entity is reloaded from the database.
```
since I don't think that we are enforcing that those references must be distinct, no ?

member must be explicitly specified.

An element collection maps to a _collection table_, as defined in
<<a14250>>.

If an embeddable class, including an embeddable class embedded
within another embeddable class, is contained, directly or
indirectly, within an element collection:

- it must not have its own element collection;
- it must not have a one-to-many or many-to-many relationship to
  an entity;
- but it may have a many-to-one or one-to-one association to an
  entity if the relationship is represented by a foreign key
  mapping, as defined in <<a516>>.

=== Map Collections [[map-collections]]

An element collection may be of type `java.util.Map`. Such a map
collection has two different element types:

- a _key type_ and
- a _value type_.

A key or value type may be a basic type, an embeddable class,
or an entity type. The key type and map type are completely
independent of one another.

A map collection must be declared using an `ElementCollection`,
`OneToMany`, or `ManyToMany` annotation:

- when the map value type is a basic type or embeddable class,
  the `ElementCollection` annotation is used;
- but when the map value type is an entity, a `OneToMany` or

>   `ManyToMany` annotation is used.

Reviewer: should we state explicitly that map key has no influence ?

Author: It's not completely necessary but if you want to, add a suggestion.

- the map maps to a join table if the map represents a many-to-many
  relationship or, by default, if it represents a unidirectional
  one-to-many relationship, but
- if the map represents the collection-valued side of a bidirectional
  one-to-many/many-to-one relationship, then, by default, the map maps
  to the primary table of the value entity type.

In this case, default mappings are described in <<a538>>.

==== Map Keys

When an element collection has declared type `Map<K,V>`, the type
argument `A` is taken, by default, to be its key type. The key type
may be explicitly specified using the `MapKeyClass` annotation. If
an element collection is declared with a raw type, the `MapKeyClass`
annotation is required.

When the map value type is an entity type, the `MapKey` annotation
specifies that the map key is itself the primary key or a persistent
field or property of the map value entity.

The `MapKey` and `MapKeyClass` annotations are mutually exclusive;
an element collection may have either a `@MapKey` or `@MapKeyClass`
annotation, or neither, but not both.

Dedicated annotations are provided for specifying column mappings
for map keys:

> - If the map key type is a basic type, the `MapKeyColumn` annotation

Reviewer: e.g. here we are introducing the annotation so I'd probably go with the fqcn.. but I'll leave it up to you to decide how you prefer to deal with the annotations 🙂

is _not_ annotated with one of the mapping annotations listed above in
<<categories>>, and if no equivalent mapping information is supplied in
any XML descriptor, the following defaulting rules are applied in order:

1. If the type of the field or property is a class annotated with the
   `Embeddable` annotation, the field or property is treated as if it
   were annotated with the `Embedded` annotation, as defined in <<a14634>>
   and <<a14672>>.
2. Otherwise, if the type of the field or property is one of the basic
   types listed in <<a486>>, it is treated as if it were annotated as
   `Basic`, as defined in <<a14205>>, <<a14719>>, <<a15087>>, and <<a16361>>.

If no annotation is present and neither of the above rules is applicable,
the behavior is undefined. A persistence provider might report an error
when the persistence unit is initialized.

=== Entity Relationships [[a516]]

Every relationship (or association) between two entity types has
well-defined _multiplicity_, either:

- **one-to-one**,
- **one-to-many** / **many-to-one** (depending on which side the relationship
  is viewed from), or
- **many-to-many**.

> An attribute which represents one side of such an association must

Reviewer: ```suggestion
An attribute that represents one side of such an association must
```

=== Inheritance [[a765]]

> An entity class might inherit another entity class. If this is the case,

Author: Also, "inherit" has "directly or indirectly" built in. "Extend" doesn't, precisely because it collides with the Java keyword.

then:

- any association targeting the second entity class is by nature a
  polymorphic association, and may reference an instance of the first
  entity class, and
- any Jakarta Persistence query against the second entity class is also

>   polymorphic and may return an instance of the first entity class.

Reviewer: maybe we should use extended/super (or similar) rather than first/second (it's not 100% clear who's first and second are here 🙂)?

Author: See [2e42aff](https://github.com/jakartaee/persistence/commit/2e42affc226b73b298655f2d9b1e909c53b85e16).

> An abstract class may be annotated `Entity` or declared as an `entity`

Reviewer: ```suggestion
An abstract class may be annotated with `Entity` or declared as an `entity`
```
?

in its XML descriptor. An abstract entity class may participate in a
relationship, and may occur in the `from` clause of a query.

An entity might extend a non-entity class, and a non-entity class might

> extend an entity class.

Reviewer: any expectations/limitations when we do so? (maybe worth pointing to some further sections?)

Reviewer: ahh ok, maybe pull the next paragraph into this one?

==== Mapped Superclasses

> An entity might inherit a superclass that declares persistent entity state and

Author: inherit (I don't mean "directly extend")

mapping information, but which is not itself an entity. Typically, the purpose
of such a _mapped superclass_ is to define state and mapping information that
is common across multiple entity classes but which maps to disparate, unrelated
database tables.

A mapped superclass must be annotated with the `MappedSuperclass` annotation or
declares as a `mapped-superclass` in its XML descriptor.

A mapped superclass might be an abstract class or a concrete class.

A mapped superclass, unlike an entity, is not queryable and must not be passed
as an argument to `EntityManager`, `EntityAgent`, or `Query` operations.

> Persistent relationships involving by a mapped superclass must be unidirectional.

Reviewer: ```suggestion
Persistent relationships involving a mapped superclass must be unidirectional.
```

=== Inheritance Mapping Strategies [[a966]]

The mapping of entity class hierarchies to database tables is specified
through metadata.

There are three basic strategies which may be used when mapping an
entity class hierarchy to tables in a relational database.

* With a **single table per class hierarchy**, every field declared by
  any class in the hierarchy maps to a column of the same table.
  A subclass shares the table of its superclasses.
* With a **table per concrete entity class**, every field declared or
  inherited by a given subclass is mapped to the same table, but each
  concrete class in the hierarchy has its own separate table.
* For the **joined subclass** strategy, any field which is declared by
  a subclass is mapped to a separate table from the fields which are
  declared by its superclass. Table joins may be used to retrieve all
  fields declared and inherited by the subclass.

These mapping strategies are enumerated by `jakarta.persistence.InheritanceType`,
and the mapping strategy for an entity class hierarchy may be specified by

> the `Inheritance` annotation or `inheritance` element of an XML descriptor.

Author: I guess we're missing a link to a14891.
